Add support for builtin DComp swapchains

n1ght-hunter opened the pull request:
**Connections**
https://github.com/gfx-rs/wgpu/issues/3486
https://github.com/gfx-rs/wgpu/issues/7108
https://github.com/gfx-rs/wgpu/issues/5150

**Description**
add support for builtin DComp swapchains to the dx12 backend.
My main goal for this was to get window transparency to work on windows

**Testing**
Tested using a basic wgpu app with transparency 

**Squash or Rebase?**
 Squash

<!--
Thanks for filing! The codeowners file will automatically request reviews from the appropriate teams.

After you get a review and have addressed any comments, please explicitly re-request a review from the
person(s) who reviewed your changes. This will make sure it gets re-added to their review queue - you're not bothering us!
-->

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `taplo format`.
- [x] Run `cargo clippy --tests`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
- [x] Run `cargo xtask test` to run tests.
- [x] If this contains user-facing changes, add a `CHANGELOG.md` entry. <!-- See instructions at the top of `CHANGELOG.md`. -->


I dont know if its normal or not but when running `cargo xtask test` i had 201 tests fail with `code 0xc0000005: Invalid access to memory location. (os error 998)` this is run on windows 10

n1ght-hunter committed:
add dx12 dcomp option to backend options

n1ght-hunter committed:
impl dcomp for the dx12 backend

n1ght-hunter committed:
Merge branch 'gfx-rs:trunk' into dev/windows-os-transparent

n1ght-hunter committed:
Merge branch 'gfx-rs:trunk' into dev/windows-os-transparent

n1ght-hunter committed:
Add Dcomp support to DX12 backend in changelog

cwfitzgerald commented on wgpu-hal/src/dx12/mod.rs:
```diff
@@ -1275,6 +1290,128 @@ impl crate::Surface for Surface {
+                                            "IDCompositionDevice::CreateVisual",
+                                        )
+                                    })?
+                                };
+
+                                {
+                                    profiling::scope!("IDCompositionTarget::SetRoot");
+                                    unsafe { target.SetRoot(&visual) }.map_err(|err| {
+                                        log::error!("IDCompositionTarget::SetRoot failed: {err}");
+                                        crate::SurfaceError::Other("IDCompositionTarget::SetRoot")
+                                    })?;
+                                }
+
+                                dcomp_state.insert(DCompState {
+                                    visual,
+                                    device: dcomp_device,
+                                    _target: target,
+                                })
+                            }
+                        };
+                        // Set the new swap chain as the content for the backing visual
+                        // and commit the changes to the composition visual tree.
+                        {
+                            profiling::scope!("IDCompositionVisual::SetContent");
+                            unsafe { dcomp_state.visual.SetContent(&swap_chain1) }.map_err(
+                                |err| {
+                                    log::error!("IDCompositionVisual::SetContent failed: {err}");
+                                    crate::SurfaceError::Other("IDCompositionVisual::SetContent")
+                                },
+                            )?;
+                        }
+
+                        // Commit the changes to the composition device.
+                        {
+                            profiling::scope!("IDCompositionDevice::Commit");
+                            unsafe { dcomp_state.device.Commit() }.map_err(|err| {
+                                log::error!("IDCompositionDevice::Commit failed: {err}");
+                                crate::SurfaceError::Other("IDCompositionDevice::Commit")
+                            })?;
+                        }
```
Could we contain all of this dcomp code into a `dcomp` module, so that this is cleaner.

cwfitzgerald commented on wgpu-types/src/instance.rs:
```diff
@@ -309,6 +309,8 @@ impl GlBackendOptions {
 pub struct Dx12BackendOptions {
     /// Which DX12 shader compiler to use.
     pub shader_compiler: Dx12Compiler,
+    /// Whether to use DirectComposition for managing presentation.
+    pub use_dcomp: bool,
```
I think we should have an enum here (which will also allow us to have a `from_env`). We could call it `Dx12PresentationSystem` and have `DXGI` and `DirectComposition` as options

cwfitzgerald commented on wgpu-hal/src/dx12/mod.rs:
```diff
@@ -527,9 +528,21 @@ struct SwapChain {
+    /// `handle` is borrowed, lifetime externally managed
+    VisualFromWndHandle {
+        handle: Foundation::HWND,
+        dcomp_state: RwLock<Option<DCompState>>,
```
I think if we architect this (in a separate module) with a struct that looks like

```rust
#[derive(Default)]
struct DCompState {
    inner: RwLock<Option<InnerState>>,
}

impl DCompState {
    // Methods for init and deinit and such
}
```

This can help keep the separation of concerns pretty clear and have this code not need to think too hard about the dcomp integration.

cwfitzgerald commented on wgpu-hal/src/dx12/mod.rs:
```diff
@@ -527,9 +528,21 @@ struct SwapChain {
+        handle: Foundation::HWND,
+        dcomp_state: RwLock<Option<DCompState>>,
+    },
     Visual(DirectComposition::IDCompositionVisual),
```
Should we be combining these enums?

n1ght-hunter replied:
honestly i dont know enough about it to say. i think what maybe would make sense is to have a Visual Enum one variant that just has Dcomp Visual and one that has the handle and other state needed

cwfitzgerald reviewed (changes requested):
Generally looking good, some comments.

n1ght-hunter commented:
will get to this in the next day or two

n1ght-hunter committed:
move decomp to another module

n1ght-hunter committed:
refactor: replace use_dcomp with presentation_system in Dx12BackendOptions

n1ght-hunter committed:
refactor: enhance Dx12PresentationSystem with Copy, PartialEq, and Eq traits

n1ght-hunter committed:
refactor: replace use_dcomp with presentation_system in Dx12 backend options

n1ght-hunter committed:
Merge branch 'trunk' into dev/windows-os-transparent

n1ght-hunter committed:
refactor: update Dx12BackendOptions to use presentation_system instead of use_dcomp

n1ght-hunter committed:
Merge branch 'dev/windows-os-transparent' of https://github.com/n1ght-hunter/wgpu into dev/windows-os-transparent

n1ght-hunter committed:
Merge branch 'trunk' into dev/windows-os-transparent

n1ght-hunter commented:
@cwfitzgerald would you be able to re review when you get time thanks

cwfitzgerald commented on wgpu-hal/src/dx12/dcomp.rs:
```diff
@@ -0,0 +1,110 @@
+}
+
+impl DCompState {
+    pub fn get_or_init(
```
This function should also be unsafe

cwfitzgerald commented on wgpu-hal/src/dx12/dcomp.rs:
```diff
@@ -0,0 +1,110 @@
+                DirectComposition::DCompositionCreateDevice::<
+                    _,
+                    DirectComposition::IDCompositionDevice,
+                >(&dxgi_device)
```
```suggestion
        let dcomp_device: DirectComposition::IDCompositionDevice = {
            profiling::scope!("DirectComposition::DCompositionCreateDevice");
            unsafe {
                DirectComposition::DCompositionCreateDevice(&dxgi_device)
```

This feels cleaner

cwfitzgerald reviewed (changes requested):
Sorry for the delay, looks good in general with a few small comments. I also think that the hardcoded list of blending modes needs to be adjusted based on if DXGI or DComp is selected

Anutrix commented:
Might totally be unfounded and not apply here but DirectComposition on Windows has issues with NVIDIA GPU when used with multiple monitors with different and high(60+) refresh rates. This was seen on Firefox so might be specific to it.
Recent ticket: https://bugzilla.mozilla.org/show_bug.cgi?id=1968876 . Couldn't find original ticket.

cwfitzgerald commented:
I definitely wouldn't worry about it here. That issue seems to be Windows 10 specific too.

melody-rs commented:
Oh! Thank you for doing this, transparency support in wgpu has been a pretty big pain point for me in the past

n1ght-hunter commented:
Im just traveling abroad atm should be able to get to this in the next week or so

cwfitzgerald commented:
Boop :)

MarijnS95 commented:
So if I understand this correctly, normally one would create and manage a composition (device, target) externally, and pass either the `IDCompositionVisual` (where `wgpu` will insert the internally-crated `IDXGISwapChain1` as content into the visual) or a `HANDLE` to an opaque "composition surface" (which the caller should insert as the content of a visual themselves - useful for cross-process sharing) into `wgpu` surface creation. Allowing users to render into arbitrary surfaces with `wgpu` while possibly combining it with a much more complicated composition/transformation tree and surfaces filled by other sources.

This PR seems to abstract away the most basic form of that, by creating and managing that `DirectComposition` device and a single `Visual` spanning the full `Target` that wraps the window (where hopefully "someone" disables the backing redirection buffer...) so that users no longer have to deal with it?

i.e. maybe this PR should be retitled and have a better description of what it achieves. Technically "dcomp on Windows" is already supported :)

MarijnS95 commented on wgpu-hal/src/dx12/dcomp.rs:
```diff
@@ -0,0 +1,110 @@
+        let dcomp_device = {
+            profiling::scope!("DirectComposition::DCompositionCreateDevice");
+            unsafe {
+                DirectComposition::DCompositionCreateDevice::<
```
As long as `Create(Virtual)Surface()` or `BeginDraw()` is not used, I'm quite sure you can just call `DCompositionCreateDevice2(NULL)` (`None` in Rust), without having to use the entire `Direct3D11on12` emulation/mapping layer.

https://learn.microsoft.com/en-us/windows/win32/api/dcomp/nf-dcomp-dcompositioncreatedevice2#remarks:

> If the renderingDevice parameter is NULL then the returned DirectComposition device cannot directly create DirectComposition surface objects. In particular, IDCompositionDevice2::CreateSurface and IDCompositionDevice2::CreateVirtualSurface methods return E_INVALIDARG, regardless of the supplied parameters. However, such a DirectComposition device object can still be used to indirectly create surfaces if the application creates a surface factory object via the IDCompositionDevice2::CreateSurfaceFactory method.

n1ght-hunter replied:
thanks i will have a look

n1ght-hunter replied:
awesome this does work. btw what is the actual difference between `DCompositionCreateDevice` and `DCompositionCreateDevice2`

n1ght-hunter commented:
sorry ended up being very busy. still traveling :). i will be home in about 3 weeks. should have time to work on it then. maybe before but unlikely. if someone else wants to fix it up no worries from me otherwise i will when i get time.

n1ght-hunter committed:
refactor(dx12): update DirectComposition device creation to use DCompositionCreateDevice2

n1ght-hunter committed:
Merge branch 'trunk' into dev/windows-os-transparent

n1ght-hunter committed:
cleanup dcomp create device types

n1ght-hunter committed:
clean dcomp code

n1ght-hunter committed:
refactor(dx12): remove unused Direct3D 11 dependencies from Cargo.toml

n1ght-hunter committed:
refactor(dx12): remove unnecessary device parameter from DirectComposition initialization

n1ght-hunter commented:
i have updated this pr now. i believe i have fixed all comments except for title of pr.
im terrible with coming up with titles so suggestions are welcome

cwfitzgerald commented on CHANGELOG.md:
````diff
@@ -42,6 +42,16 @@ Bottom level categories:
+```diff
+-pub struct Dx12BackendOptions { pub shader_compiler: Dx12Compiler }
++pub struct Dx12BackendOptions { pub shader_compiler: Dx12Compiler, pub presentation_system: Dx12PresentationSystem }
+```
````
````suggestion
#### Builtin Support for DirectComposition/DXGI swapchains in DX12

By selecting DirectComposition, the compositor cannot optimize your Swapchain as much, but you can support transparent windows.

```diff
-pub struct Dx12BackendOptions { pub shader_compiler: Dx12Compiler }
+pub struct Dx12BackendOptions { pub shader_compiler: Dx12Compiler, pub presentation_system: Dx12PresentationSystem }
```

By @n1ght-hunter in [#7550](https://github.com/gfx-rs/wgpu/pull/7550).
````

cwfitzgerald commented on CHANGELOG.md:
````diff
@@ -42,6 +42,16 @@ Bottom level categories:
++pub struct Dx12BackendOptions { pub shader_compiler: Dx12Compiler, pub presentation_system: Dx12PresentationSystem }
+```
+
+#### General
````
The next section is actually also supposed to be in Major Changes
```suggestion
```

cwfitzgerald commented on wgpu-types/src/instance.rs:
```diff
@@ -429,6 +437,47 @@ impl NoopBackendOptions {
+    #[default]
+    Dxgi,
+    /// Use the DirectComposition presentation system.
+    Dcomp,
```
```suggestion
    DirectComposition,
```

I should have mentioned this earlier, sorry, lets use the full name here.

cwfitzgerald commented on wgpu-types/src/instance.rs:
```diff
@@ -429,6 +437,47 @@ impl NoopBackendOptions {
+    /// Use the DXGI presentation system.
+    #[default]
+    Dxgi,
+    /// Use the DirectComposition presentation system.
```
```suggestion
    /// Use a DXGI swapchain made from a DirectComposition visual made from the window.
    ///
    /// This supports transparent windows, but does not support fullscreen optimization.
```

MarijnS95 replied:
@cwfitzgerald these user-facing documentation bits need links to upstream documentation exactly how all this works. Perhaps we need to explain that users cannot or should not use this in conjunction with `SurfaceTargetUnsafe::CompositionVisual`/`SurfaceTargetUnsafe::SurfaceHandle` if they're already managing a layer tree externally (themselves).

---

Can you backup (link) exactly how this doesn't support "fullscreen optimization" (like borderless fullscreen)? I was under the impression that this uses flip-style present as well which should allow for hardware-compositable/presentable planes in the sawpchain?

EDIT: Or are you saying this because translucent surfaces inevitably require the underlying plane(s) to be rendered and blended (hopefully [accelerated in hardware by MPO blending](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/d3dumddi/ne-d3dumddi-_d3dddi_multiplane_overlay_blend)) too? If the user disables alpha blending on the swapchain it should still be able to make this optimization.

MarijnS95 replied:
Should we also recommend users to set `WS_EX_NOREDIRECTIONBITMAP`?

cwfitzgerald replied:
> these user-facing documentation bits need links to upstream documentation exactly how all this works. Perhaps we need to explain that users cannot or should not use this in conjunction with SurfaceTargetUnsafe::CompositionVisual/SurfaceTargetUnsafe::SurfaceHandle if they're already managing a layer tree externally (themselves).

I agree, though we can leave this to another PR and spin off an issue from this one, I don't want to hold it up too long as I've already let it sit a while.

> Can you backup (link) exactly how this doesn't support "fullscreen optimization" (like borderless fullscreen)? I was under the impression that this uses flip-style present as well which should allow for hardware-compositable/presentable planes in the sawpchain?

I thought I had a conversation about this with Jesse on the d3d12 discord a while ago but I can't find it, so I'll ask again and see what they say.

> Should we also recommend users to set WS_EX_NOREDIRECTIONBITMAP?

I honestly don't know.

cwfitzgerald replied:
Re: mpo

I was wrong! Things made from visuals are still eligible for full screen optimization.

cwfitzgerald commented on wgpu-types/src/instance.rs:
```diff
@@ -429,6 +437,47 @@ impl NoopBackendOptions {
+#[derive(Clone, Debug, Default, Copy, PartialEq, Eq)]
+/// Selects which presentation system to use on DX12.
+pub enum Dx12PresentationSystem {
+    /// Use the DXGI presentation system.
```
```suggestion
    /// Use a DXGI swapchain made directly from the window's HWND.
    ///
    /// This supports fullscreen optimization, making borderless windows just as efficient as exclusive fullscreen. It does not support transparent windows.
```

cwfitzgerald commented on wgpu-types/src/instance.rs:
```diff
@@ -429,6 +437,47 @@ impl NoopBackendOptions {
 }
 
+#[derive(Clone, Debug, Default, Copy, PartialEq, Eq)]
+/// Selects which presentation system to use on DX12.
+pub enum Dx12PresentationSystem {
```
```suggestion
/// Selects which kind of swapchain to use on DX12.
pub enum Dx12SwapchainKind {
```

Sorry for another rename, but I think this is more accurate.

cwfitzgerald reviewed (changes requested):
Sorry for another round of comments, we're almost there!

n1ght-hunter committed:
Update wgpu-types/src/instance.rs

Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

n1ght-hunter committed:
Update wgpu-types/src/instance.rs

Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

n1ght-hunter committed:
Update wgpu-types/src/instance.rs

Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

n1ght-hunter committed:
Update wgpu-types/src/instance.rs

Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

n1ght-hunter committed:
Update CHANGELOG.md

Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

n1ght-hunter committed:
Update CHANGELOG.md

Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

n1ght-hunter committed:
Merge branch 'trunk' into dev/windows-os-transparent

n1ght-hunter committed:
fmt

n1ght-hunter commented:
@cwfitzgerald done :)

MarijnS95 commented on wgpu-types/src/instance.rs:
```diff
@@ -439,6 +444,51 @@ impl NoopBackendOptions {
+            .as_deref()?
+            .to_lowercase();
+        match value.as_str() {
+            "dcomp" => Some(Self::DirectComposition),
```
Also update the argument name?

n1ght-hunter replied:
oh ya was gonna ask we wanted to update that. though i could see leaving it so its short in the cli

cwfitzgerald replied:
I think having both would be reasonable as well.

MarijnS95 commented on wgpu-hal/src/dx12/dcomp.rs:
```diff
@@ -0,0 +1,70 @@
+            profiling::scope!("DirectComposition::DCompositionCreateDevice");
+            unsafe { DirectComposition::DCompositionCreateDevice2(None) }.map_err(|err| {
+                log::error!("DirectComposition::DCompositionCreateDevice failed: {err}");
+                crate::SurfaceError::Other("DirectComposition::DCompositionCreateDevice")
```
It's surprising that `SurfaceError` doesn't have an inner error like `InstanceError`.

n1ght-hunter replied:
whats the go with? i presume that is for another pr if required?

MarijnS95 commented on CHANGELOG.md:
```diff
@@ -42,6 +42,17 @@ Bottom level categories:
 
+#### Builtin Support for DirectComposition/DXGI swapchains in DX12
+
+By selecting DirectComposition, the compositor cannot optimize your Swapchain as much, but you can support transparent windows.
```
```suggestion
By selecting DirectComposition, the compositor cannot optimize your Swapchain as much, but you can support transparent windows.

This creates a single `IDCompositionVisual` over the entire window that is used by the `Surface`. If a user wants to manage the composition tree themselves, they should create their own device and composition, and pass the relevant visual down into `wgpu` via `IDCompositionTarget::CompositionVisual` or `IDCompositionTarget::SurfaceHandle`.
```

MarijnS95 commented on wgpu-hal/src/dx12/dcomp.rs:
```diff
@@ -0,0 +1,70 @@
+    /// Creates a DirectComposition device and a target for the given window handle.
+    pub unsafe fn init(hwnd: &HWND) -> Result<Self, crate::SurfaceError> {
+        let dcomp_device: DirectComposition::IDCompositionDevice = {
+            profiling::scope!("DirectComposition::DCompositionCreateDevice");
```
Just wondering, is it normal to wrap every external API call - especially these that are well off the hot path - in a profiling scope?

n1ght-hunter replied:
perhaps just a single profile scope for the whole init scope?

cwfitzgerald replied:
Yeah lets do a single scope - generally the decision for what to scope depends on how long it takes, so if there's a single item that takes a long time, scope that, but for processes that only really take a while in aggregate, we can use a single scope.

melody-rs commented:
Could we maybe get an example showing how to use DComp to make window transparency work? 

Window transparency with wgpu has been a frustration for a while now and having a working example to base things off of would be super helpful imo

n1ght-hunter commented:
> Could we maybe get an example showing how to use DComp to make window transparency work?
> 
> Window transparency with wgpu has been a frustration for a while now and having a working example to base things off of would be super helpful imo

sure i have a testing example i can its basiclly just a triangle with partly clear background

n1ght-hunter committed:
Update CHANGELOG.md

Co-authored-by: Marijn Suijten <marijns95@gmail.com>

cwfitzgerald commented:
> Could we maybe get an example showing how to use DComp to make window transparency work?

I think we should, but we can do this as a follow up.

cwfitzgerald commented:
@n1ght-hunter casual poke!

n1ght-hunter committed:
use a single scope profle for all of DCompState init

n1ght-hunter committed:
Allow "directcomposition" as an alias for "dcomp" in Dx12SwapchainKind

n1ght-hunter committed:
Merge branch 'trunk' into dev/windows-os-transparent

cwfitzgerald reviewed (approved):
I have a few final tweaks, but I'll push a commit to do them so we don't need to do another round of this.

cwfitzgerald committed:
Final tweeks

cwfitzgerald committed:
merge

MarijnS95 commented on CHANGELOG.md:
```diff
@@ -42,15 +42,23 @@ Bottom level categories:
+By enabling DirectComposition support, the dx12 backend can now support transparent windows.
 
-This creates a single `IDCompositionVisual` over the entire window that is used by the `Surface`. If a user wants to manage the composition tree themselves, they should create their own device and composition, and pass the relevant visual down into `wgpu` via `IDCompositionTarget::CompositionVisual` or `IDCompositionTarget::SurfaceHandle`.
+This creates a single `IDCompositionVisual` over the entire window that is used by the mf`Surface`. If a user wants to manage the composition tree themselves, they should create their own device and composition, and pass the relevant visual down into `wgpu` via `SurfaceTargetUnsafe::CompositionVisual`.
```
@cwfitzgerald what is an `` mf`Surface` ``?

Also ``SurfaceTargetUnsafe::SurfaceHandle`` is still an option, especially when shared across processes.

cwfitzgerald replied:
> what is an mf`Surface`?

A typo 😅 

Thanks for catching these issues, I'll push a PR fixing these up.

MarijnS95 replied:
Sure, feel free to ping me if you'd like a review of that followup PR :)

MarijnS95 commented on wgpu-types/src/instance.rs:
```diff
@@ -449,29 +449,36 @@ impl NoopBackendOptions {
     /// Valid values, case insensitive:
-    /// - `Dxgi`
-    /// - `DirectComposition`
+    /// - `DxgiFromVisual` or `Visual`
+    /// - `DxgiFromHwnd` or `Hwnd` for [`Self::DxgiFromHwnd`]
```
Does the first line miss ``for [`Self::DxgiFromVisual`]``, to match the `Hwnd` line?

MarijnS95 commented on wgpu-types/src/instance.rs:
```diff
@@ -449,29 +449,36 @@ impl NoopBackendOptions {
-    /// This supports transparent windows, but does not support fullscreen optimization.
-    DirectComposition,
+    /// This creates a single [`IDCompositionVisual`] over the entire window that is used by the `Surface`.
+    /// If a user wants to manage the composition tree themselves, they should create their own device and
```
I hope it's clear to users that you don't mean to create a _graphics_ device but an _`IDCompositionDevice`_ :sweat_smile:

MarijnS95 commented on wgpu-types/src/instance.rs:
```diff
@@ -449,29 +449,36 @@ impl NoopBackendOptions {
     /// Use a DXGI swapchain made directly from the window's HWND.
     ///
-    /// This supports fullscreen optimization, making borderless windows just as efficient as exclusive fullscreen. It does not support transparent windows.
+    /// This does not support transparency but has better support from developer tooling from RenderDoc.
```
from developer tooling _like_ RenderDoc?
